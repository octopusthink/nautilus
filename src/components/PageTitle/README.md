A Page Title is a piece of text that describes the current page or screen's content.

You can think of `<PageTitle>` as a combination `<title>` and `<h1>` tag; it is used to describe the content of the entire page/screen available to the user, and should not change unless the user navigates away from the current view. **Every page you create should have one and only one `<PageTitle>` component.**

```jsx
<PageTitle>My Homepage</PageTitle>
```

## Usage

Use a `<PageTitle>` to introduce the user to the entirety of the content available in the current view. On a blog, for instance, your index view might use: `<PageTitle>My Posts</PageTitle>` and an individual post would use its post title: `<PageTitle>{post.title}</PageTitle>`.

<!-- TODO for @tofumatt: implement this feature. -->
When using `<DocumentTitle>`, which uses `react-helmet`, the content inside `<PageTitle>` will be inserted into your page's `<title>` tag. If your `<PageTitle>` contains a link or other content aside from plain text, you should use the `documentTitle` prop to set the content that will be used by `<DocumentTitle>`:

```jsx inside Markdown
const title = 'This title is linked';
<PageTitle documentTitle={title}>
  <a href="/#/ðŸ›  Function/Components/PageTitle">{title}</a>
</PageTitle>
```

Don't use a `<PageTitle>` for:
- Text describing only part of a page's content. Use `<Heading>` instead.
- Creating multiple `<h1>` components. There should only be one `<PageTitle>` on the page.

## Appearance

- TODO for @sarahmonster.

## Variations

TODO for @sarahmonster.

## Voice & Tone

A page title should ideally be very short and describe the content available to the user on the page. If the page is for editorial content (eg. a blog post or other piece of prose), the title of said prose is appropriate (even if it's long). If the page is for an app, something short and descriptive like `<PageTitle>Manage Account</PageTitle>` is appropriate. Avoiding describing nested content or actions; `<PageTitle>Manage Account: Update your Credit Card Information and Change your Email</PageTitle>` would be better split into separate `<Heading>` sections for each action.

## Accessibility

For accessibility reasons, it's important that `<PageTitle>` exists to describe the purpose of the page to a user of assistive technology. It's also useful for search engine crawlers.

- Always use a single `<PageTitle>` tag per page.
