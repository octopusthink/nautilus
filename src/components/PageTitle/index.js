import { css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React from 'react';

import { pageTitle } from 'themes/mixins';

export const PageTitle = ({ children, documentTitle, ...otherProps }) => {
  // const documentTitleToUse = documentTitle || children;

  return <h1 {...otherProps}>{children}</h1>;
};

PageTitle.defaultProps = {
  children: undefined,
  documentTitle: undefined,
};

PageTitle.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** String to pass to the `<DocumentTitle />` tag if the children of your `<PageTitle />` are more than just text content. */
  documentTitle: PropTypes.string,
};

export default styled(PageTitle)(({ theme }) => {
  return css`
    ${pageTitle(theme)}
  `;
});
