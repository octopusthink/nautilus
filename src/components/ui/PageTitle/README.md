A Page Title is a short line of text that describes the current page or screen. Every page or screen in your site or app should have **one** `<PageTitle>` component: no more, no less.

```jsx
<PageTitle noMargin>Super-groovy fantastico homepage!</PageTitle>
```

## Usage

Use a `<PageTitle>` to provide a visual indication and summary of the current page or screen.

Don't use a `<PageTitle>` for:
- Text describing only a portion of a page's content. Use `<Heading>` instead.
- Creating multiple `<h1>` components. There should only be one `<PageTitle>` on the page.

## Technical Implementation

You can think of `<PageTitle>` as a combination `<title>` and `<h1>` tag. It is used to describe the content of the entire page/screen available to the user, and should not change unless the user navigates away from the current view. **Every page you create** should have **one and only one** `<PageTitle>` component.

Use a `<PageTitle>` to introduce the user to the entirety of the content available in the current view. On a blog, for instance, your index view might use: `<PageTitle>My Posts</PageTitle>` and an individual post would use its post title: `<PageTitle>{post.title}</PageTitle>`.

 When using `<DocumentTitle>`, which uses `react-helmet`, the content inside `<PageTitle>` will be inserted into your page's `<title>` tag. If your `<PageTitle>` contains a link or other content aside from plain text, you should use the `documentTitle` prop to set the content that will be used by `<DocumentTitle>`:

 ```jsx inside Markdown
 const title = 'This title is linked';
 <PageTitle noMargin documentTitle={title}>
   <a href="/#/ðŸ›  Function/Components/PageTitle">{title}</a>
 </PageTitle>
 ```

## Appearance

- Use a consistent style rules for `<PageTitle>`s across your site, even if you're doing creative art direction.
- Since a `<PageTitle>` is highest in hierarchy, it should be larger or have more emphasis than other `<Heading>`s.
- `<PageTitle>`s can use styling similar to `<Heading>`s, but doesn't need to. Since they are the highest in the hierarchy and only appear once per view, you can feel free to get a little creative with them. Highlight your brand, apply funky styles like gradients or skews, use hand-lettering effects, or art-direct each one with different typography and special styling.

## Voice & Tone

- Keep the title short and descriptive of the page's content.
- `<PageTitle>` can use Sentence case or Title Case, depending on your personal tastes and the formality of your site or app. Whichever you pick, be consistent throughout!
- If the page is for editorial content (eg. a blog post or other piece of prose), the title of said prose is appropriate (even if it's long).
- If the page is for an app, something short and descriptive like `Manage Account` is appropriate.
- Avoiding describing nested content or actions. `Manage Account: Update your Credit Card Information and Change your Email` would be better split into separate `<Heading>` sections for each action.

## Accessibility

For accessibility reasons, it's important that `<PageTitle>` exists to describe the purpose of the page to a user of assistive technology. It's also useful for search engine crawlers.

- Every page or screen in your site or app should have **one** `<PageTitle>` component.
- Don't use more than **one** `<PageTitle>` component on a single page or screen.
- Your `<PageTitle>` should *only* be the title of your site on your homepage. On all other pages, the `<PageTitle>` should describe the content of that page.
