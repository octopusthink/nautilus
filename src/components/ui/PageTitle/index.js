import { css } from '@emotion/core';
import PropTypes from 'prop-types';
import React from 'react';

import { pageTitle } from 'styles';
import { toUnits } from 'styles';
import { useTheme } from 'themes';

export const PageTitle = (props) => {
  const theme = useTheme();
  const { children, documentTitle, noMargin, unstyled, ...otherProps } = props;
  // const documentTitleToUse = documentTitle || children;

  return (
    <h1
      css={
        unstyled
          ? undefined
          : css`
              ${pageTitle(theme)}
              margin: 0 0 ${toUnits(theme.spacing.margin.large)};

              ${noMargin &&
                css`
                  margin: 0;
                `}
            `
      }
      {...otherProps}
    >
      {children}
    </h1>
  );
};

PageTitle.defaultProps = {
  children: undefined,
  documentTitle: undefined,
  noMargin: false,
  unstyled: false,
};

PageTitle.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** String to pass to the `<DocumentTitle />` tag if the children of your `<PageTitle />` are more than just text content. */
  documentTitle: PropTypes.string,
  /** Remove any outer margins from component. */
  noMargin: PropTypes.bool,
  /* @ignore Don't output any CSS styles. */
  unstyled: PropTypes.bool,
};

export const { defaultProps, propTypes } = PageTitle;

export default PageTitle;
