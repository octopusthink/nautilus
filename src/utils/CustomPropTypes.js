import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
