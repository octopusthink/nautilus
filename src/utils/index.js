export { default as CustomPropTypes } from './CustomPropTypes';
