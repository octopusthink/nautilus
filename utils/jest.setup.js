// import axios from 'axios';
// import AxiosMockAdapter from 'axios-mock-adapter';
// import { JSDOM } from 'jsdom';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-axe/extend-expect';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom/extend-expect';

global.window.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: jest.fn(),
  removeListener: jest.fn(),
}));

jest.spyOn(global.console, 'error');
jest.spyOn(global.console, 'warn');
jest.spyOn(global.console, 'log');
jest.spyOn(global.console, 'info');
jest.spyOn(global.console, 'debug');

beforeEach(() => {
  global.console.error.mockClear();
  global.console.warn.mockClear();
  global.console.log.mockClear();
  global.console.info.mockClear();
  global.console.debug.mockClear();
});
