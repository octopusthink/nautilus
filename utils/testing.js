// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
