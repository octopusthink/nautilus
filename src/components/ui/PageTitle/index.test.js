import React from 'react';

import { axe, render } from 'utils/testing';

import PageTitle from '.';

describe('PageTitle', () => {
  it('should render an <h1>', () => {
    const { container } = render(<PageTitle>My Account</PageTitle>);

    expect(container.firstChild.tagName).toEqual('H1');
  });

  it('should output its children', () => {
    const { getByTestId } = render(
      <PageTitle>
        <div data-testid="child" />
      </PageTitle>,
    );

    expect(getByTestId('child')).toBeDefined();
  });

  it('should output pageTitle CSS', () => {
    const { container } = render(<PageTitle>My Blog Posts</PageTitle>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not output CSS when unstyled prop is set', () => {
    const { container } = render(<PageTitle unstyled>My Blog Posts</PageTitle>);

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('accessibility', () => {
    it('should pass aXe tests', async () => {
      const { container } = render(<PageTitle>My Blog Posts</PageTitle>);

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });
  });
});
