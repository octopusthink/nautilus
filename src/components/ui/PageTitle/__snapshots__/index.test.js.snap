// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageTitle should not output CSS when unstyled prop is set 1`] = `
<h1
  class="emotion-0"
>
  My Blog Posts
</h1>
`;

exports[`PageTitle should output pageTitle CSS 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 700;
  font-size: 4.2rem;
  line-height: 1.2380952380952381;
  margin: 0;
  margin: 0 0 4.8rem;
}

<h1
  class="emotion-0"
>
  My Blog Posts
</h1>
`;

exports[`PageTitle shouldn't have a margin when \`noMargin\` is true 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 700;
  font-size: 4.2rem;
  line-height: 1.2380952380952381;
  margin: 0;
}

<h1
  class="emotion-0"
>
  Hello
</h1>
`;
